RPI5: add sdhc driver and sd-card support

GrygiriiS opened the pull request:
Add Broadcom BCM2712 SDHC driver which implements SD Host Controller v3.0.

Only removable SD-card interface is supported and tested with FAT
file-system.

Supported/tested:
- only SD Card support verified;
- bus power ON/OFF;
- only 3.3V bus voltage;
- bus width 1,4;
- clock rate 40KHz, 25MHz, 50MHz;
- timings Legacy/SDR12, SDR25, DDR50;
- signal voltage 3.3V and 1.8V with DDR50;
- PIO transfer;
- SDMA transfer;
- only polling transfer.

Not supported/tested:
- no IRQ support
- ADMA;
- no tuning;
- SDR50 can be enabled, but no tuning;
- SDR104

Tested with samples/subsys/fs/fs_sample:
  west build -b rpi_5 -p always samples/subsys/fs/fs_sample

with CONFIG_FILE_SYSTEM_SHELL enabled.

Tested with tests/drivers/sdhc:

  west build -b rpi_5 -p always tests/drivers/sdhc/

 - PASS - [sdhc.test_card_if_cond] duration = 1.503 seconds
 - PASS - [sdhc.test_card_presence] duration = 0.001 seconds
 - PASS - [sdhc.test_host_props] duration = 0.001 seconds
 - PASS - [sdhc.test_reset] duration = 0.002 seconds
 - PASS - [sdhc.test_set_io] duration = 0.008 seconds

danieldegrasse committed:
tests: drivers: sdhc: setup SDHC IO before test

Rather than relying to the alphanumeric test ordering implemented in
ZTEST to ensure certain tests run first, setup the SDHC IO properties to
known good values before starting any of the tests. This allows the SDHC
tests to run in any order, rather than needing certain tests to run
first.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
(cherry picked from commit 2be937dca442f02c7ef8e944540af5de2681dbe5)

Commit:
dts: bindings: add bcm2712 sdhc

Add DT bindings for Broadcom BCM2712 SDHC v3.0 module, which is present on
raspberry pi 5 platform.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
dts: arm64: bcm2712: pl011 add interrupt-names

The UART "arm,pl011" required interrupt-names (any values) to be defined to
enable UART IRQs.

Hence, add interrupt-names to "arm,pl011" node.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
dts: arm64: broadcom: bcm2712: add sdio1 sdhc node

Add DT node for BCM2712 SDHC sdio1 which handles removable SD-card slot on
Raspberry PI 5.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
dts: boards: arm64: rpi_5: add basic sd-card support

The Raspberry PI 5 has SD-card slot which is serviced by sdio1
mmc@1000fff000 SDHC module. This patch adds basic DT nodes required to
enabled sdmmc support on rpi_5 board.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
samples: subsys: fs: fs_sample: enable for rpi_5

Enable samples/subsys/fs/fs_sample for rpi_5 board.
- add shell
- disable SD card modifications

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
drivers: sdhc: add bcm2712 sdhci driver

Add Broadcom BCM2712 SDHC driver which implements SD Host Controller v3.0.

Only removable SD-card interface is supported and tested with FAT
file-system.

Supported/tested:
- only SD Card support verified;
- bus power ON/OFF;
- only 3.3V bus voltage;
- bus width 1,4;
- clock rate 40KHz, 25MHz, 50MHz;
- timings Legacy/SDR12, SDR25, DDR50;
- signal voltage 3.3V and 1.8V with DDR50;
- PIO transfer;
- SDMA transfer;
- only polling transfer.

Not supported/tested:
- no IRQ support
- ADMA;
- no tuning;
- SDR50 can be enabled, but no tuning;
- SDR104

Tested with samples/subsys/fs/fs_sample:
  west build -b rpi_5 -p always samples/subsys/fs/fs_sample

with CONFIG_FILE_SYSTEM_SHELL enabled.

Tested with tests/drivers/sdhc:

  west build -b rpi_5 -p always tests/drivers/sdhc/

 - PASS - [sdhc.test_card_if_cond] duration = 1.503 seconds
 - PASS - [sdhc.test_card_presence] duration = 0.001 seconds
 - PASS - [sdhc.test_host_props] duration = 0.001 seconds
 - PASS - [sdhc.test_reset] duration = 0.002 seconds
 - PASS - [sdhc.test_set_io] duration = 0.008 seconds

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

Commit:
boards: arm64: rpi_5: dts: increase ram size

RPI5 has 4G of memory and some samples/tests require more memory than it's
defined now, so increase RAM size to 4M.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

lorc commented on drivers/sdhc/CMakeLists.txt:
```diff
@@ -11,4 +11,5 @@ zephyr_library_sources_ifdef(CONFIG_SAM_HSMCI sam_hsmci.c)
 zephyr_library_sources_ifdef(CONFIG_INTEL_EMMC_HOST intel_emmc_host.c)
 zephyr_library_sources_ifdef(CONFIG_SDHC_INFINEON_CAT1 ifx_cat1_sdio.c)
 zephyr_library_sources_ifdef(CONFIG_CDNS_SDHC sdhc_cdns_ll.c sdhc_cdns.c)
+zephyr_library_sources_ifdef(CONFIG_BRCM_BCM2712_SDHCI brcm_bcm2712_sdhci.c sdhci_common.c)
```
Well, if you are implementing generic SDHCI driver, it should use separate `CONFIG_` option because other vendor's drivers might use it.

GrygiriiS replied:
fixed

lorc commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,801 @@
+	}
+
+	if (on) {
+
```
Stray empty line

GrygiriiS replied:
fixed

lorc commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,801 @@
+	uintptr_t dma_addr;
+
+	if (!is_read) {
+		sys_cache_data_flush_range(data->data, (data->blocks * data->block_size));
```
You need to flush caches in both cases: read and write. Consider the following sequence for reading:

1. Client allocates read buffer
2. Client makes `memzero(read_buffer)` -> this gets cached.
3. Client starts SDHC read operation to `read_buffer`
4. HW writes to `read_buffer`
4. **At the same time** CPU decides to flush caches and zeroes from p.2 overwrite newly read data.

GrygiriiS replied:
Huh. I can update it this way - np. But that's how most of Zephyr drivers doing it :(

GrygiriiS replied:
updated

lorc reviewed (commented):
Great job! When are you planning to upstream this?

GrygiriiS commented:
@oleksiimoisieiev @dsemenets @xakep-amatop I'd appreciated for review

lorc reviewed (approved):
Thanks. No more comments from my side

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

Commit:
fs/fatfs: Add CONFIG_FS_FATFS_LBA64

The Kconfig option enables support for 64-bit LBA, which also allows
to enable GPT support.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

GrygiriiS commented:
Updated with Kconfig option to support GPT with FATFS

Commit:
fs/fatfs: Add CONFIG_FS_MULTI_PARTITION

This option switches support for multiple volumes on the physical drive. By
default (0), each logical drive number is bound to the same physical drive
number and only an FAT volume found on the physical drive will be mounted.
When this function is enabled (1), each logical drive number can be bound
to arbitrary physical drive and partition listed in the VolToPart[].

For example, 2 FAT partition on SD disk (3) in terms of Zephyr:
   {3, 1} - mount point "/0:"
   {3, 2} - mount point "/1:"
The mount points have to be numbered in this case.

Signed-off-by: Grygorii Strashko <grygorii_strashko@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

GrygiriiS commented:
Added tags. Added Kconfig option to enable multi-partition support in FATFS module.

lorc commented:
For the last commit (`fs/fatfs: Add CONFIG_FS_MULTI_PARTITION`): 

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

xakep-amatop commented on dts/arm64/broadcom/bcm2712.dtsi:
```diff
@@ -73,6 +73,7 @@
 			reg = <0x10 0x7d001000 0x200>;
 			interrupts = <GIC_SPI 121 IRQ_TYPE_LEVEL
 				      IRQ_DEFAULT_PRIORITY>;
+			interrupt-names = "irq_121";
```
1. How is this change related to the current PR? As far as I can see, this is related to the console subsystem. Is it possible to open a separate PR for that?
2. Should we mark `interrupt-names` as a required property in `arm,pl011.yaml`?

GrygiriiS replied:
1. This change is required to run fs_sample and on this stage i'm not going to split this PR for rpi_5
2. `arm,pl011.yaml` can and work normally without IRQ, so no it is not mandatory.

xakep-amatop replied:
2. I don't know, maybe YAML files have something like 'if this property is present, then we also MUST have property_2'. In any case, it is just a suggestion, so you can ignore it.

xakep-amatop commented on drivers/regulator/regulator_fixed.c:
I see that the commit message describes changes that aren't present in the commit:
```
Fix it, by removing regulator_common_init_enable() from
regulator_fixed_init() and just regulator_enable() istead if
regulator_fixed is enabled at boot ("regulator-boot-on").
```

-----
Please fix the misspellings in the commit message too.

GrygiriiS replied:
> I see that the commit message describes changes that aren't present in the commit:

No sure I've got this comment :( I'd be appreciated if you could be more specific, please.

xakep-amatop replied:
I don't see function with name `regulator_common_init_enable`, moreover you remove `regulator_common_init`
`istead` --> `instead`

looks like you forgot to add some verb here: `and just regulator_enable() istead`

xakep-amatop commented on boards/arm64/rpi_5/rpi_5_defconfig:
```diff
@@ -10,3 +10,5 @@ CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+
+CONFIG_CACHE_MANAGEMENT=y
```
Would it be better to select it from the appropriate DMA config option in the driver Kconfig*?

xakep-amatop commented on boards/arm64/rpi_5/rpi_5.dts:
```diff
@@ -35,6 +35,11 @@
 
+/* samples required more memory and RPI5 has it, set it to 4M */
+&sram0 {
+	reg = <0x0 0x200000 0x400000>;
```
Maybe it is better to set it even higher for future use, like 1-2 GB?

GrygiriiS replied:
It's questionable - 4M is enough for default, non  XEN related use-cases.

xakep-amatop commented on include/zephyr/drivers/sdhc.h:
```diff
@@ -159,7 +159,7 @@ enum sd_voltage {
 struct sdhc_host_caps {
-	unsigned int timeout_clk_freq: 5;		/**< Timeout clock frequency */
+	unsigned int timeout_clk_freq: 6;		/**< Timeout clock frequency */
 	unsigned int _rsvd_6: 1;			/**< Reserved */
```
It looks like somebody wants to name `_rsvd_x` fields using exact positions, but as far as I can see, it is broken below before your changes.

Moreover, according to this naming, I expect that we want to have 64 bits in this structure, but as far as I can see, we have 66. I am not even taking into account the broken boundaries of the `u32` type by one of the bitfields.

xakep-amatop replied:
looks like after your changes `slot_type` stays between two u32 words, so maybe it is needed to delete `_rsvd_6`

GrygiriiS replied:
> It looks like somebody wants to name `_rsvd_x` fields using exact positions, but as far as I can see, it is broken below before your changes.
> 
> Moreover, according to this naming, I expect that we want to have 64 bits in this structure, but as far as I can see, we have 66. I am not even taking into account the broken boundaries of the `u32` type by one of the bitfields.

This structure doesn't follow exactly SDHC standard capabilities and rather specified supported features from Zephyr sdhc interface point of view. It also not mapped on HW registers directly.

oleksiimoisieiev commented:
got warnings while building on top of your branchs:
[mmc_warns.txt](https://github.com/user-attachments/files/15532445/mmc_warns.txt)

GrygiriiS commented:
> got warnings while building on top of your branchs: [mmc_warns.txt](https://github.com/user-attachments/files/15532445/mmc_warns.txt)

Regulator warning is not my problem - it's upstream issue.

What branch are you using exactly - the current code is differ:
```
static int sdhci_check_version(struct sdhci_common *sdhci_ctx)
{
	bool supported;

	sdhci_ctx->version = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_VERSION);

	supported = FIELD_GET(SDHCI_SPEC_VER_MASK, sdhci_ctx->version) >= SDHCI_SPEC_300;
	LOG_INF("sdhc: detected ver:%lu vendor_ver:%lu %s",
		FIELD_GET(SDHCI_SPEC_VER_MASK, sdhci_ctx->version),
		FIELD_GET(SDHCI_VENDOR_VER_MASK, sdhci_ctx->version),
		supported ? "supported" : "unsupported");

	if (!supported) {
		return -ENOTSUP;
	}

	sdhci_ctx->caps = sys_read32(sdhci_ctx->reg_base + SDHCI_CAPABILITIES);
	sdhci_ctx->caps1 = sys_read32(sdhci_ctx->reg_base + SDHCI_CAPABILITIES_1);

	LOG_INF("sdhc: capabilities caps0:%x caps1:%x", sdhci_ctx->caps, sdhci_ctx->caps1);

	return 0;
}
```

oleksiimoisieiev commented:
> Regulator warning is not my problem - it's upstream issue.
> 
The second warning caused by my debug code. so please ignore it. But first one looks like you have redefined ret

xakep-amatop commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -93,14 +94,17 @@ static int regulator_fixed_init(const struct device *dev)
 		}
 	}
 
-	return regulator_common_init(dev, false);
```
It looks like this issue can also be resolved if you add a get_voltage handler to the fixed regulator driver. 

Moreover, I reviewed the usage of regulator-fixed nodes across the projects and found that most often developers haven't set min/max voltage in the device tree source, and they don't catch the issue you're facing. So, the simpler fix would be to remove the min/max properties from the DTS.

xakep-amatop commented:
`Acked-by: Mykola Kvach <mykola_kvach@epam.com>` for all commits after introducing appropriate fixes (or by explaining that we don't need the fixes, along with the reason), except for e77e238caf91751c9632902e0fcde34d15ceaa19.

xakep-amatop commented:
> > Regulator warning is not my problem - it's upstream issue.
> 
> The second warning caused by my debug code. so please ignore it. But first one looks like you have redefined ret

It has been already fixed in mainline: https://github.com/zephyrproject-rtos/zephyr/pull/72759

GrygiriiS commented:
Changes:
- dropped patch "drivers: sdhc: fix timeout_clk_freq bits width"
- dropped patch "drivers: regulator: fixed: fix init to not call". Instead dropped DT properties 		`regulator-min-microvolt = <3300000>; 		regulator-max-microvolt = <3300000>;` for `vmmc` regulator from bard file which also solves an issue.
- dropped "boards: arm64: rpi_5: enable CONFIG_CACHE_MANAGEMENT". Instead CACHE_MANAGEMENT is selected by BRCM SDHC driver

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+struct bcm2712_sdhci_config {
+	DEVICE_MMIO_NAMED_ROM(host);
+	DEVICE_MMIO_NAMED_ROM(cfg);
+	DEVICE_MMIO_NAMED_ROM(busisol);
```
Do we need this and below fields and their mappings below if they are unused?

GrygiriiS replied:
Could be dropped if you insist.

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+
+			k_msleep(cfg->power_delay_ms);
+
+			sdhci_set_bus_power(&data->sdhci_ctx, SD_VOL_3_3_V);
```
should we pass `ios->signal_voltage` here instead of `SD_VOL_3_3_V`?
or maybe we need to move this call under the next condition?

GrygiriiS replied:
No. There is a Bus power and signal voltage. And this place is about Bus power on this platform only 3.3V is supported. This call enables Bus power.

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+	}
+
+	if (host_io->timing != ios->timing) {
+		ret = sdhci_set_uhs_timing(&data->sdhci_ctx, ios->timing);
```
as for me this function has a little bit strange name, io]t can set timings not only for `ultra high speed (uhs)` modes

GrygiriiS replied:
The  SDHC parameter name "UHS Mode Select" the input parameter name "timing". Not sure what's strange. Also  it follows uboot/linux notation.

xakep-amatop replied:
Reading the name of this common function, I, as a reviewer of the code, thought that it is designed for UHS timing modes only because the name suggests that. But what about High Speed and Normal timing modes?

For instance, n the SDHCI init function, we have the following line:
```
host_caps->high_spd_support = sdhci_ctx->caps & SDHCI_CAN_DO_HISPD ? 1 : 0;
```

Okay, I see that you use `sdhci_set_uhs_timing` only for setting UHS modes, but according to the SD Host Controller Simplified Specification version 3.0 and even according to your SDHCI init code, it is possible to have non-UHS timing modes (look at section 3.9 'Changing Bus Speed Mode' from the aforementioned specification).

GrygiriiS replied:
ok. I created TODO "sdhc: rename sdhci_set_uhs_timing to sdhci_set_timing"

xakep-amatop replied:
or you can create an issue for our Zephyr repo with link to this comment

GrygiriiS replied:
https://github.com/orgs/xen-troops/projects/6/views/1?pane=issue&itemId=69912481

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+	.get_card_present = bcm2712_sdhci_get_card_present,
+	.reset = bcm2712_sdhci_reset,
+	.card_busy = bcm2712_sdhci_card_busy,
+	.execute_tuning = bcm2712_execute_tuning,
```
We may drop this function entirely as we do nothing inside.

GrygiriiS replied:
No. We can't. It's required for only one reason to allow DDR50 and 1.8V signal switch - for which Zephyr requests tuning, but as per SDHC standard tuning for DDR50 is not required. There are appropriate comments in bcm2712_execute_tuning()

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+	props->is_spi = 0;
+	props->power_delay = cfg->power_delay_ms;
+	host_caps->bus_4_bit_support = cfg->bw_4bit;
+	host_caps->hs200_support = cfg->hs200_mode;
```
Why can't we store data from DTS directly inside `host_caps->hs200_support`? I mean, why not omit the usage of `cfg->hs200_mode`?

The same question applies to other fields in the device configuration that are used only for copying values into the data device structure.

GrygiriiS replied:
It can be. Personally I follow this style (widely used in other driver) in all my drivers  to strictly split RO vs RW data and initializers.

xakep-amatop replied:
But do you need such an approach? You have never reset the host properties and have never used this RO value except in this place. So, as for me, you may perform the initialization inside BCM2712_SDHCI_INIT.

GrygiriiS replied:
Note. that when you doing such things the `data` will be moved from bss 2 data section. 
And this is question of opinion and coding style which i prefer and which is pretty common for Zephyr -
 so, with all do respect, I'll not satisfy this request.

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+{
+	struct bcm2712_sdhci_data *data = dev->data;
+
+	return sdhci_reset(&data->sdhci_ctx, SDHCI_RESET_ALL);
```
Does a reset clear only errors, or does it also reset IO settings?
If it does, should we restore IO settings after the reset?

xakep-amatop commented on drivers/sdhc/brcm_bcm2712_sdhci.c:
```diff
@@ -0,0 +1,484 @@
+	sys_write32(reg, cfg_base + SDIO_CFG_MAX_50MHZ_MODE);
+
+	/* Guesstimate the timer frequency (controller base clock) */
+	base_clk_mhz = cfg->clk_freq / (1000 * 1000);
```
Can `clk_freq` be below 1 MHz?

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+		if (!WAIT_FOR(((sys_read16(sdhci_ctx->reg_base + SDHCI_CLOCK_CONTROL) &
+				SDHCI_CLOCK_INT_STABLE)),
+			      timeout, k_msleep(1))) {
+			LOG_ERR("sdhc:clk_en: timeout for clk stable");
```
Should we continue processing in this case? I mean, should we enable the clock below?

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+		}
+	}
+
+	if (on) {
```
why can't we use above check and just add else case for it ?

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+		ctrl |= SDHCI_CTRL_4BITBUS;
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		ctrl |= SDHCI_CTRL_8BITBUS;
```
what if `SDHCI_CTRL_4BITBUS` bit is still set ?

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	switch (timing) {
+	case SDHC_TIMING_LEGACY:
+	case SDHC_TIMING_SDR12:
+		reg |= SDHCI_CTRL_UHS_SDR12;
```
as far as I understand in case if we use SDR12 + 3.3V signaling this is not UHS
so, as for me naming of defines is a little bit strange

xakep-amatop replied:
I see that  `SDHCI_HOST_CONTROL2` has only UHS timing bits

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	hdc_cmd.opcode = SD_STOP_TRANSMISSION;
+	hdc_cmd.arg = 0;
+	hdc_cmd.response_type = SD_RSP_TYPE_R1;
+	hdc_cmd.timeout_ms = 1000;
```
maybe we can use `SD_CMD_TIMEOUT` here and set it to necessary value

GrygiriiS replied:
yep.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	uint32_t *offs = (uint32_t *)data;
+	int i;
+
+	for (i = 0; i < size; i += 4, offs++) {
```
Please correct me if I'm wrong, but in cases where we have a size unaligned to a u32 word, can we read/write some garbage or should we add an assert in cases where the size isn't aligned on u32?

GrygiriiS replied:
This is configured by SDHC_BUFFER_ALIGNMENT. buffers are always u32 aligned (or SDHC_BUFFER_ALIGNMENT) aligned

xakep-amatop replied:
I'm talking about the size of the buffer. `SDHC_BUFFER_ALIGNMENT` means that we align the starting address of the buffer on some boundary.

GrygiriiS replied:
xfer always happens in blocks, so **always** u32 aligned.

xakep-amatop replied:
but block size can be 1, 2 bytes )
in any case, SH host control spec v3 describes in details how to access buffer register in this case

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	if (!long_resp) {
+		return;
+	}
+	cmd->response[1] = sys_read32(sdhci_ctx->reg_base + SDHCI_RESPONSE + 0x4);
```
we can use `SDHCI_RESPONSE1`, `SDHCI_RESPONSE2` and `SDHCI_RESPONSE3` here

GrygiriiS replied:
yep.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+
+	while (timeout > 0) {
+		reg_normal_int_stat = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (reg_normal_int_stat & SDHCI_INT_XFER_COMPLETE) {
```
This function is called for both DMA and non-DMA modes. Should we check another flag here, such as `SDHCI_INT_DMA_END`, or even `SDHCI_INT_DATA_MASK`, in case an error occurs earlier than the timeout expiration?

GrygiriiS replied:
No. No DMA events is expected here - DMA is a specific event which is not used by current implementation.
Error checked below, but yeah may be checked in loop also.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	uint8_t *data_ptr = data->data;
+	int32_t timeout;
+
+	timeout = data->timeout_ms ? data->timeout_ms : 1;
```
looks like timeout_ms for data and cmd can be negative and be equal to `SDHC_TIMEOUT_FOREVER`

GrygiriiS replied:
FOREVER not supported for polling - plan was to support for IRQ.

xakep-amatop replied:
Why not? Is there a timeout error in this case returned by host controller, and do you leave the loop?

GrygiriiS replied:
Created "TODO: consider proper handling of FOREVER timeout in poling mode"

xakep-amatop commented:
The review is finished, and I believe I found a few issues. Please fix them if you agree, or explain why not.
Unfortunately, the review is poor because I don't have the hardware manual for the registers of the RPI 5 board.

xakep-amatop commented on drivers/sdhc/sdhci_common.c:
```diff
@@ -0,0 +1,794 @@
+	reg &= ~SDHCI_CTRL_UHS_MASK;
+
+	switch (timing) {
+	case SDHC_TIMING_LEGACY:
```
you trying use UHS-I timing bits for setting `legacy mode`. Pls, review Host Control 2 register description (2.2.24 Host Control 2 Register (Offset 03Eh)) from `SD Host Controller Simplified Specification Version 3.00`

xakep-amatop commented on boards/arm64/rpi_5/rpi_5.dts:
```diff
@@ -33,6 +35,32 @@
 		};
 	};
+
+	sd_io_1v8_reg: sd_io_1v8_reg {
```
why does the name contain `1v8` if it can be configured to 3.3V?

GrygiriiS replied:
The DT is not invented by me self - it's this way in RPI5 Kernel. I think that, if you wish, you can comment in their project.

xakep-amatop replied:
You requested a review from me, and I asked a question regarding this. I don't see any indication that this code isn't yours. Moreover, you signed off on this code.

If you are not the author and don't understand what is going on here, should you add the original author and license in the header of the file and to the appropriate commit message?

GrygiriiS replied:
i'm not going to comment in this particular thread.

xakep-amatop reviewed (commented):
@GrygiriiS, some of the questions are unanswered from the previous review, and I have new ones. Please fix them or give an answer as to why we don't need to fix them.
